include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,1289 @@
+	{
+		result = eval_and_pdf_t(bxdf_eval_t(0.0), pdf);
+	}
+	else
+	{
+		mat2x4 srcs = instr_fetchSrcRegs(instr, regs);
+		BEGIN_CASES(op)
+#ifdef OP_DIFFUSE
+		CASE_BEGIN(op, OP_DIFFUSE) {
+			result.xyz = instr_execute_cos_eval_DIFFUSE(instr, s, params, bsdf_data);
+			result.w = pdf;
+		} CASE_END
+#endif
+#ifdef OP_DIFFTRANS
+		CASE_BEGIN(op, OP_DIFFTRANS) {
+			result.xyz = instr_execute_cos_eval_DIFFTRANS(instr, s, params, bsdf_data);
+			result.w = pdf;
+		} CASE_END
+#endif
+#ifdef OP_CONDUCTOR
+		CASE_BEGIN(op, OP_CONDUCTOR) {
+			result.xyz = instr_execute_cos_eval_CONDUCTOR(instr, ior, s, microfacet, bxdf_eval_scalar_part, params, bsdf_data);
+			result.w = pdf;
+		} CASE_END
+#endif
+#ifdef OP_COATING
+		CASE_BEGIN(op, OP_COATING) {
+			//(in instr_t instr, in mat2x4 srcs, in irr_glsl_LightSample s, in irr_glsl_AnisotropicMicrofacetCache microfacet, in float DG, in params_t params, in bsdf_data_t data, in float coat_pdf)
+			result = instr_execute_cos_eval_pdf_COATING(instr, srcs, params, ior[0], ior2[0], s, microfacet, bxdf_eval_scalar_part, bsdf_data, pdf);
+		} CASE_END
+#endif
+#ifdef OP_DIELECTRIC
+		CASE_BEGIN(op, OP_DIELECTRIC) {
+			result.xyz = instr_execute_cos_eval_DIELECTRIC(instr, s, bxdf_eval_scalar_part);
+			result.w = pdf;
+		} CASE_END
+#endif
+#ifdef OP_THINDIELECTRIC
+		CASE_BEGIN(op, OP_THINDIELECTRIC) {
+			result = instr_execute_cos_eval_pdf_THINDIELECTRIC(instr, s, regs, params, bsdf_data);
+		} CASE_END
+#endif
+#ifdef OP_BLEND
+		CASE_BEGIN(op, OP_BLEND) {
+			result = instr_execute_cos_eval_pdf_BLEND(instr, srcs, params, bsdf_data);
+		} CASE_END
+#endif
+#ifdef OP_BUMPMAP
+		CASE_BEGIN(op, OP_BUMPMAP) {
+			instr_execute_BUMPMAP(instr, srcs, precomp);
+		} CASE_END
+#endif
+#ifdef OP_SET_GEOM_NORMAL
+		CASE_BEGIN(op, OP_SET_GEOM_NORMAL) {
+			instr_execute_SET_GEOM_NORMAL(instr, precomp);
+		} CASE_END
+#endif
+		CASE_OTHERWISE
+		{} //else "empty braces"
+		END_CASES
+	}
+
+	if (op_isBXDForBlend(op))
+		writeReg(REG_DST(regs), result);
+}
+
+eval_and_pdf_t irr_bsdf_eval_and_pdf(in MC_precomputed_t precomp, in instr_stream_t stream, inout irr_glsl_LightSample s, in uint generator_offset, inout irr_glsl_AnisotropicMicrofacetCache microfacet)
+{
+	setCurrInteraction(precomp);
+	for (uint i = 0u; i < stream.count; ++i)
+	{
+		instr_t instr = irr_glsl_MC_fetchInstr(stream.offset+i);
+		uint op = instr_getOpcode(instr);
+
+		if (i != generator_offset)
+		{
+			instr_eval_and_pdf_execute(instr, precomp, s, microfacet);
+		}
+		else
+		{
+			//assert(isBXDF(instr));
+			uvec3 regs = instr_decodeRegisters(instr);
+			writeReg(REG_DST(regs), eval_and_pdf_t(0.0));
+		}
+
+#if defined(OP_SET_GEOM_NORMAL)||defined(OP_BUMPMAP)
+		if (
+#ifdef OP_SET_GEOM_NORMAL
+			op == OP_SET_GEOM_NORMAL
+#ifdef OP_BUMPMAP
+			||
+#endif
+#endif
+#ifdef OP_BUMPMAP
+			op == OP_BUMPMAP
+#endif
+			) {
+			s = irr_glsl_createLightSample(s.L, currInteraction);
+			microfacet = irr_glsl_calcAnisotropicMicrofacetCache(currInteraction, s);
```

Comment:
why recompute from scratch? if the View and Light positions don't change then your half vector doesnt change and its dot product with V or L don't either